src/io/flutter/editor/NativeEditorNotificationProvider.java

```diff
@@ -0,0 +1,138 @@
+ * found in the LICENSE file.
+ */
+package io.flutter.editor;
+
+import com.intellij.openapi.actionSystem.*;
+import com.intellij.openapi.editor.colors.EditorColors;
+import com.intellij.openapi.fileEditor.FileEditor;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Key;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.ui.EditorNotificationPanel;
+import com.intellij.ui.EditorNotifications;
+import com.intellij.ui.HyperlinkLabel;
+import icons.FlutterIcons;
+import org.jetbrains.annotations.NonNls;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class NativeEditorNotificationProvider extends EditorNotifications.Provider<EditorNotificationPanel> implements DumbAware {
+  private static final Key<EditorNotificationPanel> KEY = Key.create("flutter.native.editor.notification");
+
+  private final Project myProject;
+
+  public NativeEditorNotificationProvider(@NotNull Project project) {
+    myProject = project;
+  }
+
+  @NotNull
+  @Override
+  public Key<EditorNotificationPanel> getKey() {
+    return KEY;
+  }
+
+  @Nullable
+  @Override
+  public EditorNotificationPanel createNotificationPanel(@NotNull VirtualFile file, @NotNull FileEditor fileEditor) {
+    if (!file.isInLocalFileSystem()) {
+      return null;
+    }
+    VirtualFile root = findRootDir(file);
+    if (root == null) {
+      return null;
+    }
+
+    String actionName;
+    if (root.getName().equals("android")) {
+      actionName = "flutter.androidstudio.open";
+    }
+    else if (root.getName().equals("ios")) {
+      actionName = "flutter.xcode.open";
+    }
+    else {
+      return null;
+    }
+
+    NativeEditorActionsPanel panel = new NativeEditorActionsPanel(file, root, actionName);
+    return panel.isValidForFile() ? panel : null;
+  }
+
+  private VirtualFile findRootDir(@NotNull VirtualFile file) {
+    // Return the top-most parent of file that is a child of the project directory.
+    VirtualFile projectDir = myProject.getBaseDir();
+    VirtualFile parent = file.getParent();
+    if (projectDir.equals(parent)) {
+      return null;
+    }
+    VirtualFile root = parent;
+    while (parent != null) {
+      parent = parent.getParent();
+      if (projectDir.equals(parent)) {
+        return root;
+      }
+      root = parent;
+    }
+    return null;
+  }
+
+  class NativeEditorActionsPanel extends EditorNotificationPanel {
+    final VirtualFile myFile;
+    final VirtualFile myRoot;
+    final AnAction myAction;
+
+    NativeEditorActionsPanel(VirtualFile file, VirtualFile root, String actionName) {
+      super(EditorColors.GUTTER_BACKGROUND);
+      myFile = file;
+      myRoot = root;
+      myAction = ActionManager.getInstance().getAction(actionName);
+
+      icon(FlutterIcons.Flutter);
+      text("Flutter commands");
```

Comment:
I wonder if we might have something more hand-holdy here; like:

"For the best results, edit ${platform} files in ${platform_tool}; open ${platform_tool}"

/cc @devoncarew @mit-mit

Reply:
We may not have room for the longer text, and the action text (`Open Android module in Android Studio`) is ~pretty descriptive.

Instead of `Flutter commands`, perhaps just `Flutter`? Between that and the flutter icon, they'll have good context for who's providing the editor action bar.

Reply:
How about 'Open for editing in Android Studio' ?

Reply:
@mit-mit That sounds like we are opening the file, which we're not (at least not yet). But if I can get that to happen then the change sounds good.

@devoncarew Should we make that same change in other places as well? I picked "Flutter commands" to be consistent.

Reply:
In the other location we are actually showing 3-4 flutter_tools commands - packages get, packages upgrade, ... So I think `Flutter commands` makes sense there.